magicgui/tqdm.py

```diff
@@ -0,0 +1,151 @@
+"""A wrapper around the tqdm.tqdm iterator that adds a ProgressBar to a magicgui."""
+import inspect
+from typing import Iterable, Optional
+
+from magicgui.application import use_app
+from magicgui.widgets import FunctionGui, ProgressBar
+
+try:
+    from tqdm import tqdm
+except ImportError as e:  # pragma: no cover
+    msg = (
+        f"{e}. To use magicgui with tqdm please `pip install tqdm`, "
+        "or use the tqdm extra: `pip install magicgui[tqdm]`"
+    )
+    raise type(e)(msg)
+
+
+def _find_calling_function_gui(max_depth=6) -> Optional[FunctionGui]:
+    """Traverse calling stack looking for a magicgui FunctionGui."""
+    for finfo in inspect.stack()[2:max_depth]:
+
+        # regardless of whether the function was decorated directly in the global module
+        # namespace, or if it was renamed on decoration (`new_name = magicgui(func)`),
+        # or if it was decorated inside of some local function scope...
+        # we will eventually hit the :meth:`FunctionGui.__call__` method, which will
+        # have the ``FunctionGui`` instance as ``self`` in its locals namespace.
+        if finfo.function == "__call__" and finfo.filename.endswith("function_gui.py"):
+            obj = finfo.frame.f_locals.get("self")
+            if isinstance(obj, FunctionGui):
+                return obj
+            return None  # pragma: no cover
+
+    return None
+
+
+_tqdm_kwargs = {
+    p.name
+    for p in inspect.signature(tqdm.__init__).parameters.values()
+    if p.kind is not inspect.Parameter.VAR_KEYWORD and p.name != "self"
+}
+
+
+class tqdm_mgui(tqdm):
```

Comment:
Curious why not just have this class be called `tqdm`, to use it you're importing `from magicgui import tqdm` already, especially if this is a drop in replacement for `tqdm` / behavior is identical if not inside a magicgui decorated function

Reply:
because it's not yet a drop-in replacement :/

Reply:
ok now it is!  I liked this idea... so now it's just `from magicgui.tqdm import tqdm, trange` ... and if it's not inside of a `@magicgui` it will work as usual (like in IPython)

Reply:
love it so much!!